src/lib/mdns/TxtFields.cpp

```diff
@@ -0,0 +1,248 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "TxtFields.h"
+
+#include <algorithm>
+#include <cctype>
+#include <cstdio>
+#include <inttypes.h>
+#include <limits>
+#include <stdlib.h>
+#include <string.h>
+
+#include <mdns/Resolver.h>
+#include <support/BytesToHex.h>
+#include <support/CHIPMemString.h>
+
+namespace chip {
+namespace Mdns {
+
+namespace Internal {
+
+namespace {
+
+char SafeToLower(uint8_t ch)
+{
+    return static_cast<char>(std::tolower(static_cast<unsigned char>(ch)));
+}
+bool IsKey(const ByteSpan & key, const char * desired)
+{
+    if (key.size() != strlen(desired))
+    {
+        return false;
+    }
+    for (size_t i = 0; i < key.size(); ++i)
+    {
+        if (SafeToLower(key.data()[i]) != SafeToLower(desired[i]))
+        {
+            return false;
+        }
+    }
+    return true;
+}
+
+uint16_t MakeU16FromAsciiDecimal(const ByteSpan & val)
+{
+    // Largest u16 number if 65536, so 5 chars.a
+    constexpr size_t kMaxUint16StrLen = 5;
+    // Add a space for null
+    char temp[kMaxUint16StrLen + 1];
+    if (val.size() > kMaxUint16StrLen)
+    {
+        return 0;
+    }
+    Platform::CopyString(temp, sizeof(temp), val);
+    char * end;
+    unsigned long num = strtoul(temp, &end, 10);
```

Comment:
This function is internally consistent in that all error paths, including those from strtoul, return 0.  However, it does not appear callers are checking this, or would even have the ability to always distinguish a successfully parsed 0 from a parsing error.

I would suggest a general interface of approximately the following, with error paths propagated to external callers:

```
CHIP_ERROR MakeU16FromAsciiDecimal(const ByteSpan & val, uint16_t &out)
```

I realize our preference is for granular PRs and commits.  But can we have a follow on task / TODO to propagate and fail on parsing errors?

Reply:
I've added an issue to double check all the fields, but the idea with using 0 was that that is the invalid value for all the supplied values.

Reply:
whoops, sorry - issue is here: https://github.com/project-chip/connectedhomeip/issues/7319